patches/server/1005-PreSpawnerSearchEvent.patch

```diff
@@ -0,0 +1,65 @@
+index 633500aefd515df5dadda3802b94079f75a03fa0..f155c4094b1c49d6ce1789518a6cb7f0962c207d 100644
+--- a/src/main/java/net/minecraft/world/level/BaseSpawner.java
++++ b/src/main/java/net/minecraft/world/level/BaseSpawner.java
+@@ -1,6 +1,7 @@
+ package net.minecraft.world.level;
+ 
+ import com.mojang.logging.LogUtils;
++
+ import java.util.Optional;
+ import java.util.function.Function;
+ import javax.annotation.Nullable;
+@@ -46,6 +47,7 @@ public abstract class BaseSpawner {
+     public int requiredPlayerRange = 16;
+     public int spawnRange = 4;
+     private int tickDelay = 0; // Paper
++    private int searchCooldownTicks = 0; // Paper
+ 
+     public BaseSpawner() {}
+ 
+@@ -57,6 +59,16 @@ public abstract class BaseSpawner {
+     public boolean isNearPlayer(Level world, BlockPos pos) {
+         return world.hasNearbyAlivePlayerThatAffectsSpawning((double) pos.getX() + 0.5D, (double) pos.getY() + 0.5D, (double) pos.getZ() + 0.5D, (double) this.requiredPlayerRange); // Paper - Affects Spawning API
+     }
++    // Paper Start - PreSpawnerSearchEvent (Convert ServerPlayers into BukkitPlayer)
++    public java.util.Collection<org.bukkit.entity.Player> convertServerPlayersToBukkitPlayers(java.util.List<net.minecraft.server.level.ServerPlayer> serverPlayers) {
++        java.util.Collection<org.bukkit.entity.Player> bukkitPlayers = new java.util.ArrayList<>(serverPlayers.size());
++
++        for (net.minecraft.server.level.ServerPlayer human : serverPlayers) {
++            bukkitPlayers.add(human.getBukkitEntity());
++        }
++        return bukkitPlayers;
++    }
++    // Paper End
+ 
+     public void clientTick(Level world, BlockPos pos) {
+         if (!this.isNearPlayer(world, pos)) {
+@@ -86,7 +98,20 @@ public abstract class BaseSpawner {
+         tickDelay = world.paperConfig().tickRates.mobSpawner;
+         if (tickDelay == -1) { return; } // If disabled
+         // Paper end
+-        if (this.isNearPlayer(world, pos)) {
++
++        // Paper start - PreSpawnerSearchEvent
++        if (--searchCooldownTicks > 0) return;
++
++        java.util.Collection<org.bukkit.entity.Player> activatingPlayers = convertServerPlayersToBukkitPlayers(world.getNearbyPlayers(null, (double) pos.getX() + 0.5D, (double) pos.getY() + 0.5D, (double) pos.getZ() + 0.5D, (double) this.requiredPlayerRange, net.minecraft.world.entity.EntitySelector.PLAYER_AFFECTS_SPAWNING)); // Uses logic from hasNearbyAlivePlayerThatAffectsSpawning
++
++        io.papermc.paper.event.entity.PreSpawnerSearchEvent preSpawnerSearchEvent = new io.papermc.paper.event.entity.PreSpawnerSearchEvent(io.papermc.paper.util.MCUtil.toLocation(world, pos), activatingPlayers, searchCooldownTicks);
++
++        if (!preSpawnerSearchEvent.callEvent()) return;
++
++        searchCooldownTicks = preSpawnerSearchEvent.getBackoffTicks();
++        // Paper end
++
++        if (!activatingPlayers.isEmpty()) { // Paper - Use PreSpawnerSearchEvent API
```

Comment:
```suggestion
+        if (!activatingPlayers.isEmpty()) { // Paper - Use PreSpawnerSearchEvent API

+            io.papermc.paper.event.entity.PreSpawnerSearchEvent preSpawnerSearchEvent = new io.papermc.paper.event.entity.PreSpawnerSearchEvent(io.papermc.paper.util.MCUtil.toLocation(world, pos), activatingPlayers, searchCooldownTicks);
+
+            if (!preSpawnerSearchEvent.callEvent()) return;
+
+            searchCooldownTicks = preSpawnerSearchEvent.getBackoffTicks();
+        // Paper end
+
```

Perhaps this is better which would make activatingPlayers `@NotNull` within the event/api

Reply:
i don't think this list was null even before your change? an empty list is not the same as a null object

Reply:
The list is never null though? I'm somewhat confused regarding this comment.

Reply:
what i mean is that even before this commit (https://github.com/PaperMC/Paper/pull/9614/commits/dd7964fcbe11e05c7e3099e2f8443aaa3e8a649f), the list was empty at best and that your change in this commit goes against your point a and c (can't exclude players in the check if the check is done before the event).